src/epic/distances/epic.py

```diff
@@ -0,0 +1,176 @@
+        coverage_sampler: Optional[samplers.BaseSampler[samplers.CoverageSample]] = None,
+    ):
+        """Initialize the EPIC distance.
+
+        Args:
+            state_sampler: The sampler for the state distribution.
+            action_sampler: The sampler for the action distribution.
+            coverage_sampler: The sampler for the coverage distribution. If not given,
+                a default sampler is constructed as drawing from the product
+                distribution induced by the distributions of state and action.
+            discount_factor: The discount factor.
+        """
+        coverage_sampler = coverage_sampler or samplers.ProductDistrCoverageSampler(action_sampler, state_sampler)
+        super().__init__(state_sampler, action_sampler, coverage_sampler, discount_factor)
+
+    def canonicalize(
+        self,
+        reward_function: types.RewardFunction,
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalize a reward function.
+
+        Applies the canonically shaped reward transformation defined in 4.1 of
+        https://arxiv.org/pdf/2006.13900.pdf.
+
+        Args:
+            reward_function: The reward function to canonicalize.
+            n_samples_cov: The number of samples to draw from the coverage distribution.
+            n_samples_can: The number of samples to draw for the canonicalization step
+                for each sample of the coverage distribution. The total number of
+                samples drawn is ``n_samples_cov * n_samples_can``.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        def canonical_reward_fn(state, action, next_state, done, /):
+            """Canonical reward function.
+            ``state``, ``action``, ``next_state`` correspond to s, a, s' in the paper
+            and ``state_sample``, ``action_sample``, ``next_state_sample``
+            correspond to S, A, S' in the paper.
+
+            Args:
+                state: The batch of state samples from the coverage distribution.
+                action: The batch of action samples from the coverage distribution.
+                next_state: The batch of next state samples from the coverage distribution.
+                done: The batch of done samples from the coverage distribution.
+
+            Returns:
+                The canonicalized reward function.
+            """
+            n_samples_cov = state.shape[0]
+
+            assert n_samples_cov == action.shape[0] == next_state.shape[0] == done.shape[0]
+
+            n_samples = n_samples_cov * n_samples_can
+
+            # Copy each sample in n_samples_cov to n_samples_can times.
+            state, action, next_state, done = utils.broadcast(state, action, next_state, done, n_samples_can)
+
+            # Create n_samples_cov * n_samples_can samples.
```

Comment:
FWIW, it's not obvious to me this is better than using all the canonicalization samples for each coverage sample. I'd imagine if the cost of sampling is low compared to computing the means, the current version makes sense, but if sampling is expensive, you'd rather just use all the samples each time?